### PR TITLE
[Seq] Proposal for dialect and `reg` op

### PR DESCRIPTION
**Pull request by teqdruid**
Notion of a register.

### CODE REVIEW

File: docs/RationaleSeq.md
Comment (lattner) [inline docs/RationaleSeq.md:20]: Since the world refers to a clocked latch as a latch, would it make sense to call this an 'unclocked latch' for precision and to avoid confusion?

File: docs/RationaleSeq.md
Comment (lattner) [inline docs/RationaleSeq.md:29]: Just a thought reading this but "one of these things is not like the other".  Register is a much higher abstraction level with a lot of design flexibility and freedom.  I'm not sure that handling it in the same way as the other better-defined and better-constrained elements is the obvious way to go.

File: docs/RationaleSeq.md
Comment (lattner) [inline docs/RationaleSeq.md:47]: I'm not an expert here, but this doesn't seem like a very general model.  We have multiported rams and many other things that exist in the world with special constraints.  Defining a common abstraction for this will be tricky.  I'd recommend nailing down the details of the other ops first.

**Review by lattner — COMMENTED** [review]
Awesome, thank you for pushing this forward.  I look forward to talking about it in a few hours.

File: docs/RationaleSeq.md
Comment (fabianschuiki) [reply docs/RationaleSeq.md:29]: Another thought: the semantics of a D-latch are fairly distinct from a D-flip-flop, and generally you cannot simply substitute one for the other (since the D-latch becomes transparent, breaking the timing isolation "contract" that flip-flops provide). My intuition would be that you always want to specify whether the storage element you are talking about in your circuit has this timing-arc-breaking behaviour of an FF, or this transparent nature of a latch. Not being able to do so could be too vague to describe anything (e.g., you wouldn't be able to describe a state machine or counter).

File: docs/RationaleSeq.md
Comment (fabianschuiki) [reply docs/RationaleSeq.md:20]: I would also suggest to treat D-Latch and D-Flip-Flop similarly in that they have a clock and accept a value into storage. All the other latch types (SR, JK, toggle, etc.) accept a kind-of strobe signal that changes the output value, but not in the sense of "copying in" an input value. Also I think in practice these are very rare and could be represented e.g. as `seq.sr_latch`. But I could see going for `seq.ff` and `seq.latch` to refer to the D-Latch and D-Flip-Flop as useful.

File: docs/RationaleSeq.md
Comment (teqdruid) [reply docs/RationaleSeq.md:29]: To address your two comments Fabian: my understanding is that a gated SR latch and D latch are functionally equivalent (modulo the data signaling) as they both have an enable (strobe) signal which makes them go transparent. A D-FF is technically the same as a d-latch, (modulo the enable vs clock labeling) but most of the time when people say "flip flop", they mean "edge-triggered flip flop".

You're correct that replacing (edge-triggered) FFs with (clocked) latches would not work in the general case, but with careful timing analysis you can. So you can implement a state machine with latches so long as you have a sufficiently short clock high period, either as a result of a high frequency or duty cycle.

Either way, I think we want to model things at a higher level wherein we don't have to reason about timing then do the timing analysis and device selection at a lower-level. In most cases (I think), one just wants to reason about registers in the way `always_ff` does and let the synthesizer select FF vs latch and worry about the timing. Anything beyond that is an optimization to top of pure digital sequential logic, IMO.

**Comment by teqdruid** [comment]
I've updated this document to my understanding of what came out of the meeting, though I converted the resetValue to an operand rather than an attribute.

File: docs/RationaleSeq.md
Comment (mwachs5) [inline docs/RationaleSeq.md:59]: So does this mean it is not possible to model a register without any reset? This seems like a bad design choice, you seem to be baking in the assumption that this will lower to always_ff, which is in conflict with the earlier statement that is only one option

File: docs/RationaleSeq.md
Comment (mwachs5) [inline docs/RationaleSeq.md:74]: What about reset type? Are you assuming asynchronous reset always? That seems like a reasonable assumption (since synchronous reset can be pulled into the logic) unless you require a reset value

File: docs/RationaleSeq.md
Comment (mwachs5) [inline docs/RationaleSeq.md:83]: It does seem some tools are persnickety about how you code up a negative edge clock. Certainly some have not been able to map clock resources correctly if the inversion is completely outside of the always statement, and others complain if there is logic inside the always @(...)

File: docs/RationaleSeq.md
Comment (mwachs5) [inline docs/RationaleSeq.md:85]: Curious if you are thinking to propagate the reset type from the signal connected to the reg.reset

File: docs/RationaleSeq.md
Comment (mwachs5) [reply docs/RationaleSeq.md:59]: Ah, I think I missed the clarification above that this is definitely `Optional`. I think that you could capture some of the discussion mentioned in the PR comments in the rationale about why it is `Optional`

Comment (teqdruid) [reply docs/RationaleSeq.md]: We didn't discuss its optionality, just whether resetValue should be an attribute or an operand. I think we all just assumed that not all registers needed to be reset, thus reset (and resetValue) should be optional.

File: docs/RationaleSeq.md
Comment (teqdruid) [reply docs/RationaleSeq.md]: I don't think it's important to specify a reset style at this level. Since this is the purely digital sequential logic layer, we don't want to have to take timing-dependent behavior into account.

File: docs/RationaleSeq.md
Comment (teqdruid) [reply docs/RationaleSeq.md:94]: Yeah, this gets into our discussion about how we will probably have to support multiple flavors of Verilog output. **sigh**

File: docs/RationaleSeq.md
Comment (teqdruid) [reply docs/RationaleSeq.md]: The idea behind `clock` and `reset` types is that we'd be able to store information about their uses -- posedge/negedge, async/sync. This is -- apparently -- useful for generating the clock/reset trees. It would also allow us to merely check that the clock/reset types match before doing some merging operation -- like creating a shift register from a series of regs. (Not that we couldn't do that by checking some attributes, it'd just be more code.)

File: docs/RationaleSeq.md
Comment (fabianschuiki) [reply docs/RationaleSeq.md]: > To address your two comments Fabian: my understanding is that a gated SR latch and D latch are functionally equivalent (modulo the data signaling) as they both have an enable (strobe) signal which makes them go transparent. A D-FF is technically the same as a d-latch, (modulo the enable vs clock labeling) but most of the time when people say "flip flop", they mean "edge-triggered flip flop".

Just nitpicking: I don't think an SR latch can become transparent (you can set or reset it with the S or R strobe). You probably can do something like that with a gated SR latch when you invert the S or R input and short them together, but then you've essentially built a D Latch ;-) (https://en.wikibooks.org/wiki/Digital_Circuits/Latches#Gated_SR_latch)

> You're correct that replacing (edge-triggered) FFs with (clocked) latches would not work in the general case, but with careful timing analysis you can. So you can implement a state machine with latches so long as you have a sufficiently short clock high period, either as a result of a high frequency or duty cycle.

Fair point! My mind was too fixed on this purely-FF / conservative timing use case.

> Either way, I think we want to model things at a higher level wherein we don't have to reason about timing then do the timing analysis and device selection at a lower-level. In most cases (I think), one just wants to reason about registers which are assumed to break timing arcs and then let the synthesizer select FF vs latch and worry about the timing. Anything beyond that is an optimization on top of pure digital sequential logic, IMO.

Fully agreed, that makes a lot of sense. In that case opting for a latch in the IR would probably stem from the user explicitly asking for one, and in general we'd just go for an abstract register, right?

Comment (teqdruid) [reply docs/RationaleSeq.md]: > > To address your two comments Fabian: my understanding is that a gated SR latch and D latch are functionally equivalent (modulo the data signaling) as they both have an enable (strobe) signal which makes them go transparent. A D-FF is technically the same as a d-latch, (modulo the enable vs clock labeling) but most of the time when people say "flip flop", they mean "edge-triggered flip flop".
> 
> Just nitpicking: I don't think an SR latch can become transparent (you can set or reset it with the S or R strobe). You probably can do something like that with a gated SR latch when you invert the S or R input and short them together, but then you've essentially built a D Latch 😉 (https://en.wikibooks.org/wiki/Digital_Circuits/Latches#Gated_SR_latch)

Actually, if you're talking about a dual-rail encoding, then it does become "transparent". 🤪


> Fully agreed, that makes a lot of sense. In that case opting for a latch in the IR would probably stem from the user explicitly asking for one, and in general we'd just go for an abstract register, right?

Yep. Or this reg op could lower to a latch op.

**Comment by lattner** [comment]
This looks good to me @teqdruid, but it is out of my technical depth on the issues and tradeoffs involved.  :-)

**Comment by teqdruid** [comment]
@lattner I don't want to merge this until I have at least one approval... Could someone who has the technical expertise involved give me approval? Ideally, I'd have two approvals. Remember, I'm not shooting for the end-all-be-all, just a solid starting point.

**Review by fabianschuiki — APPROVED** [review]
Sorry about the delay @teqdruid, I think this is a great starting point!

File: docs/RationaleSeq.md
Comment (darthscsi) [inline docs/RationaleSeq.md:34]: Maybe say no intrinsic sub-addressing?  Setting/reading is as a whole.

File: docs/RationaleSeq.md
Comment (darthscsi) [inline docs/RationaleSeq.md:52]: include pos/neg option so we can model negedge flipflops?

File: docs/RationaleSeq.md
Comment (darthscsi) [inline docs/RationaleSeq.md:58]: [, %reset, %resetValue] per the text above

File: docs/RationaleSeq.md
Comment (darthscsi) [inline docs/RationaleSeq.md:53]: like clock, do we include an edge specifier?  is sync/async in the type or in the op?

File: docs/RationaleSeq.md
Comment (darthscsi) [inline docs/RationaleSeq.md:87]: We do need to know async/sync for codegen, so is this op encoded or type encoded?

File: docs/RationaleSeq.md
Comment (darthscsi) [inline docs/RationaleSeq.md:82]: This seems like it might lead to problems, but going with the simpler thing here (as written) seems a good start.

File: docs/RationaleSeq.md
Comment (darthscsi) [inline docs/RationaleSeq.md:100]: Our currently supported front-end language, firrtl, uses macros and stuff for initial values and not 'x.  To support this front end, we have to have custom initialization.

**Review by darthscsi — COMMENTED** [review]
I would propose inital value be supported off the bat, and we decide on how reset type is encoded (I would suggest the op attribute, rather than type based)

File: docs/RationaleSeq.md
Comment (seldridge) [reply docs/RationaleSeq.md:100]: Can you clarify on what is `x` here? In Verilog this could be any of: (1) uninitialized, (2) a one-time random value, (3) a random value every time you read, (4) the user "doesn't care" what the value is and the compiler can choose anything, (5) verification poison, or something else.

Put differently, how would this compose if the target language is VHDL where you have separeate representations for things like the above? I would like to keep the `seq` and `comb` dialects relatively un-opinionated and not wind up repeating the mistakes of Verilog with imprecision around `x`.

One solution would be to make the initial value an actual MLIR value which can support a connection from another dialect, e.g., this could support a connection from a VHDL constant (unititialized, `U`, or don't care, `-`) or a Verilog `x`.

File: docs/RationaleSeq.md
Comment (teqdruid) [reply docs/RationaleSeq.md:34]: I thought that was implied, but I can add that note.

File: docs/RationaleSeq.md
Comment (teqdruid) [reply docs/RationaleSeq.md:52]: I would rather not even specify the posedge but given that `i1` is a wire, I kinda have to and I think `posedge` is a reasonable start. I think we should cross this bridge whenever we have the discussion about clock and reset types.

File: docs/RationaleSeq.md
Comment (teqdruid) [reply docs/RationaleSeq.md:53]: I don't think modeling this is necessary at this level of abstraction and/or at this stage. I want to see how far this very simple op gets us.

File: docs/RationaleSeq.md
Comment (teqdruid) [reply docs/RationaleSeq.md:87]: IMO, this is a style thing. For any given CIRCT design, one _generally_ wants the same reset type. If we find that's not the case we can talk about modeling reset types.

File: docs/RationaleSeq.md
Comment (teqdruid) [reply docs/RationaleSeq.md:100]: I've kinda been resisting having an initial value attribute since I think it'll affect sequential optimizations. I'd rather blanket set it to (1) uninitialized so as to not preclude any optimizations which need to do any number of things which affect the initial value.

I'll change 'x to uninitialized to clarify to start. If we find in the common case an initial value is necessary, we can add it as an optional attribute.

File: docs/RationaleSeq.md
Comment (mwachs5) [reply docs/RationaleSeq.md:87]: i don't think that's a good assumption to make.

Comment (teqdruid) [reply docs/RationaleSeq.md:87]: Why? HLS workloads certainly won't care. I can't find a place in the FIRRTL IR where one can specific async vs sync reset.

Comment (mwachs5) [reply docs/RationaleSeq.md:87]: In FIRRTL IR it is inferred from the reset signal itself (which is reset or async). I'm not saying that's how it should be done, but it certainly "knows" at the time of the emission. If that's how you want to do it also I suggest adding that to the rationale doc.

personally I might want some justification why we want to represent sync reset at all, since in theory that could always be pulled back into the logic. Not sure sync reset flops are a real hardware thing.

Comment (seldridge) [reply docs/RationaleSeq.md:87]: There is some documentation on how this works here: https://www.chisel-lang.org/chisel3/docs/wiki-deprecated/reset.html. tl;dr: Three types of reset (which are represented in FIRRTL IR):

Comment (teqdruid) [reply docs/RationaleSeq.md:87]: > In FIRRTL IR it is inferred from the reset signal itself (which is reset or async typed). I'm not saying that's how it should be done, but it certainly "knows" at the time of the emission. If that's how you want to do it also I suggest adding that to the rationale doc.

I'm not sure if that's the way we want to go or not -- we have yet to discuss type based vs attribute based for clock and reset styles. Also bear in mind that this will **not** be directly emitted by ExportVerilog. It will be lowered into another dialect first by any lowering pass which can decide on the actual implementation.

> NOTE: by "FIRRTL IR" I am referring to the scala implementation, which I realized might not be what you were referring to.

I was just quickly skimming the FIRRTL dialect code. By FIRRTL IR, I was referring to the FIRRTL dialect. @lattner / @darthscsi Are reset types currently supported?

> personally I might want some justification why we want to represent sync reset at all, since in theory that could always be pulled back into the logic. Not sure sync reset flops are a real hardware thing.

Not to be flippant, but nothing besides doped sand is an actual hardware thing. After that, it's all just levels of abstraction. This is not intended to be directly emitted to hardware. It's intended to support that sweet spot at which relatively high-level modeling is possible (e.g. LLHD should be able to compile fast and accurate simulators) and lowering to an actual implementation is possible. It should make the simple, common case easy to reason about.

Comment (teqdruid) [reply docs/RationaleSeq.md:87]: > I do note that this is one solution that we fell into. However, it's perilous to assume that all registers in a compilation unit have the same reset type.

That's part of my concern of modeling reset types too soon. We need to carefully consider the abstraction we provide at this level, which I don't think is absolutely necessary to start with.

File: docs/RationaleSeq.md
Comment (teqdruid) [reply docs/RationaleSeq.md:50]: Good catch.

File: docs/RationaleSeq.md
Comment (teqdruid) [reply docs/RationaleSeq.md:9]: I cannot immediately see the feedback loop to capture state in that diagram, so I would call that combinational.

File: docs/RationaleSeq.md
Comment (teqdruid) [reply docs/RationaleSeq.md:82]: A subsequent step would recognize the inverted clock and synthesize a negedge (if we don't have an alternative by the time ppl want to use negedge clocks). We do this in other places.

File: docs/RationaleSeq.md
Comment (teqdruid) [reply docs/RationaleSeq.md:54]: I very specifically do not encode that information here. We'll leave it to a lower level to make or validate these things.

File: docs/RationaleSeq.md
Comment (teqdruid) [reply docs/RationaleSeq.md:96]: I agree that a `ResetType` and `ClockType` makes sense, but this is something we'll have to discuss in the future. I don't think we're prepared to discuss this yet.

File: docs/RationaleSeq.md
Comment (teqdruid) [reply docs/RationaleSeq.md:52]: The term "register" is pretty overloaded. When I hear 'register' I think of something which has a clock input.

**Comment by teqdruid** [comment]
It feels like people are getting hung up on the term 'register' and `reg`, so I'm tempted to rename this op to indicate that it's a relatively high-level notion of single-value, clocked storage. Any suggestions?

File: docs/RationaleSeq.md
Comment (jackkoenig) [reply docs/RationaleSeq.md:87]: I'm pretty sure every single design needs async reset, even if the rest of the design follows sync reset. Synchronizers, clock domain crossing FIFOs, etc, so if you model sync reset as reset at all, then you need to distinguish. If the goal is to keep it as simple as possible, then I think @mwachs5's suggestion of just not modeling sync reset as reset is the best option.

**Comment by mwachs5** [comment]
> I'm tempted to rename this op to indicate that it's a relatively high-level notion of single-value, clocked storage. Any suggestions?

Personally I think the name `reg` is fine. Alternatives could be `flop` or `state`?

File: docs/RationaleSeq.md
Comment (teqdruid) [reply docs/RationaleSeq.md:87]: @fabianschuiki Are there any issues with async reset modeling in LLHD? Would it cause any performance issues with the generated simulators? (esp. for what I continue to think is the common case of either don't care or sync reset.)

File: docs/RationaleSeq.md
Comment (drom) [reply docs/RationaleSeq.md:9]: And this is equivalent circuit, that might look more "sequential"?
image
The point is, that in general, any "sequential" circuit is just "combinational" circuit with a loop.

**Comment by drom** [comment]
> It feels like people are getting hung up on the term 'register' and `reg`, so I'm tempted to rename this op to indicate that it's a relatively high-level notion of single-value, clocked storage. Any suggestions?

`ff` ? `dff` ? `Flip-Flop` ?

File: docs/RationaleSeq.md
Comment (teqdruid) [reply docs/RationaleSeq.md:87]: > I'm pretty sure every single design* needs async reset, even if the rest of the design follows a sync reset discipline. Synchronizers, clock domain crossing FIFOs, etc, so if you model sync reset as reset at all, then you need to distinguish.

I'm not arguing that this op could be use to model every use of SystemVerilog registers. Rather, I intended it to model  completely synchronous "logic support" registers. E.g. FSM regs, pipeline regs. In these cases, you generally don't care about the reset behavior from a correctness standpoint -- adding specification of the reset behavior just introduces unnecessary complexity for this use case. It's like the Chisel notion of "Reset-Agnostic Code", which I suspect is the most common case in computational cores.

CDC registers are somewhat different and I think it makes sense to model them differently at this level. They often need to be coded in very particular ways and/or accompanied by a timing constraints file, which makes them different from a purely mechanical standpoint. They also should be (I suspect) optimized differently. In the general case, optimizations shouldn't touch CDC regs as pretty much any optimization would break them.

**Comment by teqdruid** [comment]
> > I'm tempted to rename this op to indicate that it's a relatively high-level notion of single-value, clocked storage. Any suggestions?
> 
> Personally I think the name `reg` is fine. Alternatives could be `flop` or `state`?

Flop is too much like flip flop, which is lower level then I'm shooting for. I like `state`, though it's a little too broad. I think starting with a very specific name would narrow the discussion to what I intended. `SingleValueState`?

File: docs/RationaleSeq.md
Comment (teqdruid) [reply docs/RationaleSeq.md:9]: Yes, you are correct. It's generally taught as two separate things, sequential after and building on combinational when I was taught. I tend to think of them as two separate (complimentary) branches of logic design, and I think that's a common view. Are you objecting to the term "orthogonal"?

**Comment by teqdruid** [comment]
I've renamed this op to 'compreg' to better imply what the use case is. `reg` implied a broader use case than I now think is feasible for a first pass op. I've also re-written the explanation text.

**Comment by mwachs5** [comment]
What does "comp" stand for? Computational?

**Comment by teqdruid** [comment]
Yes

File: docs/RationaleSeq.md
Comment (lattner) [reply docs/RationaleSeq.md]: I'm also not clear what you're driving at.  I don't think we need to define the term "combinational circuit", we just need to define "combinational" vs "sequential" operations.

File: docs/RationaleSeq.md
Comment (mwachs5) [inline docs/RationaleSeq.md:62]: If we are considering treading the "async vs sync" behavior of the reset as a higher level concept, why even capture the reset wire, either? Or the clock? I could certainly see value in separating out the "stateful logic to do the thing you want to do" from the "clock and reset domain" considerations. As it is right now it feels like we are straddling the line in an awkward way.

Comment (teqdruid) [reply docs/RationaleSeq.md:62]: I've previously proposed the "clock/reset" regions with implicit clock and reset signals. That's too high level, but I want to add it to a higher level dialect at some point. It makes lowering to this op difficult for all of the existing dialects to lower to -- they all expect a clock operand. The reason to capture the reset wire is that not all registers have to be reset -- as a rule of thumb, only the control regs have to be.

> As it is right now it feels like we are straddling the line in an awkward way.

I'm not saying that this should be unspecified forever, I'm just saying that we shouldn't design that specification method *now*. I'd rather get in a very simple MVP and get some experience with using it / analyzing it before we decide how to model resets if they prove necessary for this particular op.

I suspect that a more general register op will be necessary (modeling different clock types and reset styles) but that this one will be useful on its own. I may be wrong about that.

If we have time tomorrow, I think we should continue discussing this in the ODM.

File: docs/RationaleSeq.md
Comment (darthscsi) [reply docs/RationaleSeq.md]: > I'm also not clear what you're driving at. I don't think we need to define the term "combinational circuit", we just need to define "combinational" vs "sequential" operations.

I think (more bike shedding on names here), seqential is not a great name, we already have "sequential" regions in the sv dialect.  Perhaps this is the "state" dialect?

Comment (lattner) [reply docs/RationaleSeq.md]: I think those are called procedural regions, but I agree with you nonetheless: these are stateful elements, "state" seems like a good name to use here.

**Comment by teqdruid** [comment]
At the Wednesday meeting, we agreed that it's more important to start somewhere and iterate than to suffer from analysis paralysis. I'm going to merge this proposal and follow it up with an implementation shortly.

**Pull request merged**
